packages/fxa-auth-client/lib/salt.ts

```
@@ -0,0 +1,173 @@
+
+ /**
+  * Checks to see if the salt is a valid V1 or V2 prefix.
+  */
+ export const VX_REGEX = new RegExp(VX_REGEX_STR);
+
+
+ /**
+  * Set of valid salt versions
+  */
+ export type SaltVersion = 1 | 2;
+
+ /**
+  * The Salt model
+  */
+ export interface Salt {
+   /**
+    * The namespace prefix
+    */
+   namespace?: string;
+
+   /**
+    * The version, which follows the names prefix
+    */
+   version?: number;
+
+   /**
+    * The salts potentially unique value.
+    *  v1 - value is the email the user initially signed up with.
+    *  v2 - value is a 16 byte 32 character random hex string created on signup or password reset.
+    */
+   value?: string;
+ }
+
+ /**
+  * Parses and validates a salt string.
+  * @param salt - A string representing the salt used for password encryption.
+  *               v1 example: identity.mozilla.com/picl/v1/quickStretch:foo@mozillal.com
+  *               v2 example: identity.mozilla.com/picl/v1/quickStretchV2:0123456789abcdef0123456789abcdef
+  * @returns A Salt model
+  */
+ export function parseSalt(salt: string): Salt {
+   if (V2_REGEX.test(salt)) {
+     return checkSalt({
+       namespace: NAMESPACE,
+       version: 2,
+       value: salt.replace(V2_REGEX, ''),
+     });
+   }
+
+   if (V1_REGEX.test(salt)) {
+     return checkSalt({
+       namespace: NAMESPACE,
+       version: 1,
+       value: salt.replace(V1_REGEX, ''),
+     });
+   }
+
+   throw new Error('invalid salt format');
+ }
+
+ /**
+  * Creates a v1 salt
+  * @param email - Users email at time up account sign up.
+  * @returns string - That has a valid salt format.
+  *
+  * For example, this could generate:
+  *   identity.mozilla.com/picl/v1/quickStretchV2:foo@mozilla.org
+  *
+  * Where the portion after the prefix is the users email at time of sign up.
+  */
+ export function createSaltV1(email: string) {
+   const salt = `${NAMESPACE}${V1_MARKER}:${email}`;
+   // Ensures salt will parse later!
+   parseSalt(salt);
```

Q: is this just ensuring the `parseSalt()` doesn't `throw`? Or should do we need the return value?

---

Yes, it's just a gate to stop anything invalid from getting created. Fail hard fail fast.